datafusion/physical-plan/src/repartition/distributor_channels.rs

```diff
@@ -188,33 +205,48 @@ impl<'a, T> Future for SendFuture<'a, T> {
         let this = &mut *self;
         assert!(this.element.is_some(), "polled ready future");
 
-        let mut guard_channel = this.channel.lock();
-
-        // receiver end still alive?
-        if !guard_channel.recv_alive {
-            return Poll::Ready(Err(SendError(
-                this.element.take().expect("just checked"),
-            )));
-        }
-
-        let mut guard_gate = this.gate.lock();
+        // lock scope
+        let to_wake = {
+            let mut guard_channel_state = this.channel.state.lock();
+
+            let Some(data) = guard_channel_state.data.as_mut() else {
+                // receiver end dead
+                return Poll::Ready(Err(SendError(
+                    this.element.take().expect("just checked"),
+                )));
+            };
+
+            // does ANY receiver need data?
+            // if so, allow sender to create another
+            if this.gate.empty_channels.load(Ordering::SeqCst) == 0 {
+                let mut guard = this.gate.send_wakers.lock();
+                if let Some(send_wakers) = guard.deref_mut() {
+                    send_wakers.push((cx.waker().clone(), this.channel.id));
+                    return Poll::Pending;
+                }
+            }
 
-        // does ANY receiver need data?
-        // if so, allow sender to create another
-        if guard_gate.empty_channels == 0 {
-            guard_gate
-                .send_wakers
-                .push((cx.waker().clone(), guard_channel.id));
-            return Poll::Pending;
-        }
+            let was_empty = data.is_empty();
+            data.push_back(this.element.take().expect("just checked"));
+
+            if was_empty {
+                this.gate.decr_empty_channels();
+
+                let to_wake = guard_channel_state
+                    .recv_wakers
+                    .as_mut()
+                    .expect("not closed");
+                let mut tmp = Vec::with_capacity(to_wake.capacity());
+                std::mem::swap(to_wake, &mut tmp);
+                tmp
```

Comment:
This might be more readable if it was in a function like

```rust
guard_channel_state.take_wakers()
```

Or something - that would also give you a place to document what was happening more